sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala

```diff
@@ -53,6 +53,64 @@ abstract class ParsedStatement extends LogicalPlan {
   final override lazy val resolved = false
 }
 
+/**
+ * Type to keep track of Hive serde info
+ */
+case class SerdeInfo(
+    storedAs: Option[String] = None,
+    formatClasses: Option[(String, String)] = None,
+    serde: Option[String] = None,
+    serdeProperties: Map[String, String] = Map.empty) {
+  // this uses assertions because validation is done in validateRowFormatFileFormat etc.
+  assert(storedAs.isEmpty || formatClasses.isEmpty,
+    s"Conflicting STORED AS $storedAs and INPUTFORMAT/OUTPUTFORMAT $formatClasses values")
+
+  def describe: String = {
+    val serdeString = if (serde.isDefined || serdeProperties.nonEmpty) {
+      "ROW FORMAT" + serde.map(sd => s" SERDE $sd").getOrElse(" DELIMITED")
+    } else {
+      ""
+    }
+
+    this match {
+      case SerdeInfo(Some(format), _, _, _) =>
+        s"STORED AS $format $serdeString"
+      case SerdeInfo(_, Some((inFormat, outFormat)), _, _) =>
+        s"INPUTFORMAT $inFormat OUTPUTFORMAT $outFormat $serdeString"
+      case _ =>
+        serdeString
+    }
+  }
+
+  def merge(other: SerdeInfo): SerdeInfo = {
+    def getOnly[T](desc: String, left: Option[T], right: Option[T]): Option[T] = {
+      (left, right) match {
+        case (Some(l), Some(r)) if l != r =>
+          assert(l == r, s"Conflicting $desc values: $l != $r")
+          left
+        case (Some(_), _) =>
+          left
+        case (_, Some(_)) =>
+          right
+        case _ =>
+          None
+      }
+    }
+
+    SerdeInfo(
+      getOnly("STORED AS", storedAs, other.storedAs),
+      getOnly("INPUTFORMAT/OUTPUTFORMAT", formatClasses, other.formatClasses),
+      getOnly("SERDE", serde, other.serde),
+      serdeProperties ++ other.serdeProperties)
```

Comment:
we should make sure only one side has serde properties, or make sure 2 serde properties have no conflict,